### PR TITLE
Fix #8866: Inline calls inserted by macros

### LINKED ISSUE #8866: Macros do not expand TASTY macros if in any form of nesting
## Minimized example

```Scala
import scala.quoted._

object OtherMacro {

  def impl(using qctx: QuoteContext): Expr[Int] =
    '{ 42 }

  inline def apply = ${ OtherMacro.impl }

}

object Macro {

  def impl(using qctx: QuoteContext): Expr[Int] = {
    import qctx.tasty._
    import util._

    // Simply returning this term is OK
    // Select.unique(
    //   '{ OtherMacro }.unseal,
    //   "apply"
    // ).seal.cast[Int]

    let(
      Select.unique(
        '{ OtherMacro }.unseal,
        "apply"
      )
    )(identity).seal.cast[Int]
  }

  inline def apply = ${ Macro.impl }

}

```

## Output

```scala
[error] 3 |  val a = Macro.apply
[error]   |          ^^^^^^^^^^^
[error]   |          method apply is declared as erased, but is in fact used
```

## Expectation

Either TASTY reflect terms shouldn't be expanded, or possibly we can expand them explicitly? If you simply return the commented out code, the macro gets expanded.

If you have a quote with a macro it gets expanded as expected.

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

**Comment by anatoliykmetyuk** [comment]
@nicolasstucki utest issue doesn't seem to do anything with utest. Here is a minimized example.

macro.scala:
```scala
import scala.quoted._

def f(xs: Boolean*): Unit = ???

def mcrImpl(using QuoteContext): Expr[Unit] =
  val func: Expr[Seq[Boolean] => Unit] =
    '{(esx: Seq[Boolean]) => f(esx: _*)}
  val trees: Expr[Seq[Boolean]] = '{Seq(true)}
  Expr.betaReduce(func)(trees)
end mcrImpl

inline def mcr: Unit = ${ mcrImpl }
```

Test.scala:
```scala
val x = mcr
```

Output:
```scala
-- [E007] Type Mismatch Error: /Users/kmetiuk/Projects/scala3/playground/nico-issue/Test.scala:1:8
1 |val x = mcr
  |        ^^^
  |        Found:    (x$1 : Seq[Boolean])
  |        Required: Boolean
  | This location contains code that was inlined from Test.scala:2
  | This location contains code that was inlined from macros.scala:7
```

Without the change introduced in this PR, the above example compiles successfully.

**Comment by nicolasstucki** [comment]
I finally got the force to finish this one.

File: compiler/src/dotty/tools/dotc/typer/Inliner.scala [hunk 1]
```diff
@@ -1239,7 +1239,8 @@ class Inliner(call: tpd.Tree, rhsToInline: tpd.Tree)(using Context) {
         case res: Apply if res.symbol == defn.InternalQuoted_exprSplice
                         && level == 0
                         && !suppressInline =>
-          expandMacro(res.args.head, tree.span)
+          val expanded = expandMacro(res.args.head, tree.span)
+          typedExpr(expanded) // Inline calls and constant fold code generated by the macro
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/typer/Inliner.scala:1243]: I remember that retype checking of leaf nodes is problematic. I was surprised that the CI is green.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/typer/Inliner.scala:1243]: I have been fixing those for a while aiming to fix this one.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/typer/Inliner.scala:1243]: Did you have any specific one mind?

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/typer/Inliner.scala:1243]: I think name resolution will be a big problem, in particular in the presence of implicits.

For example, given the following program:

```Scala
def test(using c: Int) = {
  def foo(c: Int): Int = summon[Int]
}
```

After type check the body of `foo`, we get the tree `c`. Retype check the name `c` will change semantics of the program.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/typer/Inliner.scala:1243]: The Inliner already retypes the tree without changing the initial elaboration. All implicit should already be resolved and not change in this tree.

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/typer/Inliner.scala:1243]: I see, because we are already in `ReTyper`.

**Review by liufengyun — APPROVED** [review]
LGTM

**Pull request merged**